[SPARK-7501] [Streaming] DAG visualization: show DStream operations

andrewor14 opened the pull request:
This is similar to #5999, but for streaming. Roughly 200 lines are tests.

One thing to note here is that we already do some kind of scoping thing for call sites, so this patch adds the new RDD operation scoping logic in the same place. Also, this patch adds a `try finally` block to set the relevant variables in a safer way.

@tdas @zsxwing

---

**Before**
<img src="https://cloud.githubusercontent.com/assets/2133137/7625996/d88211b8-f9b4-11e4-90b9-e11baa52d6d7.png" width="450px"/>

---

**After**
<img src="https://cloud.githubusercontent.com/assets/2133137/7625997/e0878f8c-f9b4-11e4-8df3-7dd611b13c87.png" width="650px"/>

Commit:
Scope streaming operations instead of RDD operations

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #32313 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32313/consoleFull) for   PR 6034 at commit [`a0d3263`](https://github.com/apache/spark/commit/a0d32630c7ebf026b38ea65874a005c9ff6e9537).

Commit:
Fix NPE

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #32316 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32316/consoleFull) for   PR 6034 at commit [`e85b787`](https://github.com/apache/spark/commit/e85b787b23d1844b89525e3b8c156e0850f53a48).

SparkQA commented:
**[Test build #32313 timed out](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32313/consoleFull)**     for PR 6034 at commit [`a0d3263`](https://github.com/apache/spark/commit/a0d32630c7ebf026b38ea65874a005c9ff6e9537)     after a configured wait of `150m`.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32313/
Test FAILed.

SparkQA commented:
[Test build #32316 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32316/consoleFull) for   PR 6034 at commit [`e85b787`](https://github.com/apache/spark/commit/e85b787b23d1844b89525e3b8c156e0850f53a48).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32316/
Test PASSed.

tdas commented on core/src/main/scala/org/apache/spark/rdd/RDD.scala:
```diff
@@ -1523,13 +1523,15 @@ abstract class RDD[T: ClassTag](
+          checkpointData.get.doCheckpoint()
+        } else {
+          dependencies.foreach(_.rdd.doCheckpoint())
+        }
```
This is not for this PR. The other PR still not merged in master .. isnt it?

andrewor14 replied:
oops I can't believe I accidentally merged these two PRs.

tdas commented on core/src/main/scala/org/apache/spark/rdd/RDDOperationScope.scala:
```diff
@@ -43,7 +43,7 @@ import org.apache.spark.SparkContext
 private[spark] class RDDOperationScope(
     val name: String,
-    val parent: Option[RDDOperationScope] = None) {
+    val parent: Option[RDDOperationScope] = None) extends Serializable {
```
Why was this made serializable?

tdas commented:
I think there should be a few unit tests to test this. See how the creation sites of the DStream are tested in StreamingContextSuite, test "DStream and generated RDD creation sites"

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/dstream/DStream.scala:
```diff
@@ -109,6 +109,18 @@ abstract class DStream[T: ClassTag] (
+   * Instead, every time we call `compute` we instantiate a new scope using the same name as this
+   * one. Otherwise, all RDDs ever created by this DStream will be in the same scope.
+   */
+  private val scope: Option[RDDOperationScope] = {
```
Can you store only the name in as a DStream field? I does not make sense to save this arbitrary mostly unnecessary RDDOperationScope object to be serialized with DStreams as part of DStream checkpoints.

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/dstream/DStream.scala:
```diff
@@ -321,6 +320,48 @@ abstract class DStream[T: ClassTag] (
+  private def doCompute(time: Time): Option[RDD[T]] = {
+    val scopeKey = SparkContext.RDD_SCOPE_KEY
+    val scopeNoOverrideKey = SparkContext.RDD_SCOPE_NO_OVERRIDE_KEY
+    // Set the thread-local property for call sites to this DStream's creation site
```
Can you update the comments to generalize it to both call sites and scopes?

tdas commented:
This doesnt yet set the scope for input stream (that is DStream that are directly created, not using DStream transformations). I think you need to use the withScope in places like StreamingContext.socketStream(). Example: 

```
class StreamingContext {

def socketTextStream(): DStream[String] = withScope {
   new SocketInputDStream[String](...)
}
```

This needs to be done for every subclass of `InputDStream` and `ReceiverInputDStream`

Commit:
Merge branch 'master' of github.com:apache/spark into dag-viz-streaming

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #32650 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32650/consoleFull) for   PR 6034 at commit [`c121047`](https://github.com/apache/spark/commit/c12104708cea8c7bc306f57f6ab380f7fa4a6887).

Commit:
Wrap many more methods in withScope

This covers all the places where we instantiate DStreams.

SparkQA commented:
[Test build #32650 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32650/consoleFull) for   PR 6034 at commit [`c121047`](https://github.com/apache/spark/commit/c12104708cea8c7bc306f57f6ab380f7fa4a6887).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class DoubleArrayParam(parent: Params, name: String, doc: String, isValid: Array[Double] => Boolean)`
  * `class Bucketizer(JavaTransformer, HasInputCol, HasOutputCol):`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32650/
Test PASSed.

Commit:
Merge branch 'master' of github.com:apache/spark into dag-viz-streaming

Commit:
Group RDDs by DStream operations and batches

This commit does two main things:

(1) We now display the batch information in each cluster on the UI
(2) Scopes are marked by the DStream operation and the batch

For (2), scopes previously only took into account the batch in
a particular DStream. This means the scopes are never shared
across multiple DStreams, and so we end up with many more boxes
(one for each DStream) per batch than is needed.

Commit:
Rename operations that create InputDStreams

Instead of displaying something like "createStream", we will
display something more descriptive like "kafka direct stream"
as of this commit.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #32676 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32676/consoleFull) for   PR 6034 at commit [`0e1438f`](https://github.com/apache/spark/commit/0e1438f83f0707ac863174b3d761e0fd3e85d455).

andrewor14 commented:
@zsxwing @tdas Check out the latest screenshot!! Also cc/ @JoshRosen who might be interested.

SparkQA commented:
[Test build #32676 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32676/consoleFull) for   PR 6034 at commit [`0e1438f`](https://github.com/apache/spark/commit/0e1438f83f0707ac863174b3d761e0fd3e85d455).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32676/
Test FAILed.

andrewor14 commented:
retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #32695 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32695/consoleFull) for   PR 6034 at commit [`0e1438f`](https://github.com/apache/spark/commit/0e1438f83f0707ac863174b3d761e0fd3e85d455).

SparkQA commented:
[Test build #32695 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32695/consoleFull) for   PR 6034 at commit [`0e1438f`](https://github.com/apache/spark/commit/0e1438f83f0707ac863174b3d761e0fd3e85d455).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32695/
Test FAILed.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #32703 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32703/consoleFull) for   PR 6034 at commit [`385442a`](https://github.com/apache/spark/commit/385442a439df5cb87ae1f66a3d67d9a7fd2f7787).

SparkQA commented:
**[Test build #32703 timed out](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32703/consoleFull)**     for PR 6034 at commit [`385442a`](https://github.com/apache/spark/commit/385442a439df5cb87ae1f66a3d67d9a7fd2f7787)     after a configured wait of `150m`.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32703/
Test FAILed.

Commit:
Merge branch 'master' of github.com:apache/spark into dag-viz-streaming

Commit:
Fix MIMA?

Commit:
Fix test

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #32720 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32720/consoleFull) for   PR 6034 at commit [`b3806ab`](https://github.com/apache/spark/commit/b3806aba729842bd7c5fff806c59d4820c2b430a).

SparkQA commented:
[Test build #32720 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32720/consoleFull) for   PR 6034 at commit [`b3806ab`](https://github.com/apache/spark/commit/b3806aba729842bd7c5fff806c59d4820c2b430a).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32720/
Test FAILed.

Commit:
Add tests for DStream scopes

AmplabJenkins commented:
Build triggered.

AmplabJenkins commented:
Build started.

SparkQA commented:
[Test build #32752 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32752/consoleFull) for   PR 6034 at commit [`9113183`](https://github.com/apache/spark/commit/9113183c10d78e5af357c6be07b321b0f1b3d118).

Commit:
Merge branch 'master' of github.com:apache/spark into dag-viz-streaming

Conflicts:
	streaming/src/main/scala/org/apache/spark/streaming/ui/StreamingPage.scala
	streaming/src/main/scala/org/apache/spark/streaming/ui/UIUtils.scala

Commit:
Fix tests?

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #32755 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32755/consoleFull) for   PR 6034 at commit [`e4a93ac`](https://github.com/apache/spark/commit/e4a93acf937f965b26ab6a04466a51f1bda24edb).

SparkQA commented:
[Test build #32752 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32752/consoleFull) for   PR 6034 at commit [`9113183`](https://github.com/apache/spark/commit/9113183c10d78e5af357c6be07b321b0f1b3d118).
 * This patch **fails Spark unit tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32752/
Test FAILed.

SparkQA commented:
[Test build #32755 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32755/consoleFull) for   PR 6034 at commit [`e4a93ac`](https://github.com/apache/spark/commit/e4a93acf937f965b26ab6a04466a51f1bda24edb).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32755/
Test PASSed.

tdas commented on core/src/main/scala/org/apache/spark/rdd/RDDOperationScope.scala:
```diff
@@ -43,10 +43,9 @@ import org.apache.spark.SparkContext
 @JsonPropertyOrder(Array("id", "name", "parent"))
 private[spark] class RDDOperationScope(
     val name: String,
+    val id: String = RDDOperationScope.nextScopeId().toString,
```
Super Nit: typically default params are ordered based on what is more likely to get explicitly specified. I think `parent` is more likely be specified by callers than the `id`, so should be ordered accordingly.

tdas commented on core/src/main/scala/org/apache/spark/rdd/RDDOperationScope.scala:
```diff
@@ -88,14 +87,25 @@ private[spark] object RDDOperationScope {
    * Execute the given body such that all RDDs created in this body will have the same scope.
-   * The name of the scope will be the name of the method that immediately encloses this one.
+   * The name of the scope will be the first method name in the stack trace that is not the
+   * same as this method's.
```
What do you mean by `this method's `? The name is not `withScope`?

tdas commented on core/src/main/scala/org/apache/spark/rdd/RDDOperationScope.scala:
```diff
@@ -88,14 +87,25 @@ private[spark] object RDDOperationScope {
       allowNesting: Boolean = false)(body: => T): T = {
-    val callerMethodName = Thread.currentThread.getStackTrace()(3).getMethodName
+    val stackTrace = Thread.currentThread.getStackTrace().tail // ignore "Thread#getStackTrace"
+    val ourMethodName = stackTrace(1).getMethodName
```
Which method name is this? `withScope`? If so mention that in the comments.

andrewor14 replied:
I didn't do it cause if we change the method name then we must update the comment. It's a little like duplicating code. But I can add it not a big deal

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/dstream/DStream.scala:
```diff
@@ -338,6 +370,48 @@ abstract class DStream[T: ClassTag] (
+    // thread-local properties in our SparkContext. Since this method may be called from another
+    // DStream, we need to temporarily store any old scope and creation site information to
+    // restore them later after setting our own.
+    // TODO: this won't work if multiple StreamingContexts share the same SparkContext
```
Why wont this work? Each streaming context will be calling compute in its own scheduling thread. They are not going to interfere with each other.

andrewor14 replied:
I see. OK.

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/dstream/InputDStream.scala:
```diff
@@ -49,6 +49,9 @@ abstract class InputDStream[T: ClassTag] (@transient ssc_ : StreamingContext)
   private[streaming] def name: String = s"${getClass.getSimpleName}-$id"
 
+  /** Human-friendly scope name to use in place of generic operation names (e.g. createStream). */
+  protected override val customScopeName: Option[String] = Some(s"input stream [$id]")
```
Why `protected` and not `private[streaming]` like `name` ?

andrewor14 replied:
because this is only used in `DStream` and overridden in its children. Classes outside of `DStream` don't need to access it.

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/dstream/FileInputDStream.scala:
```diff
@@ -126,6 +126,8 @@ class FileInputDStream[K, V, F <: NewInputFormat[K,V]](
   @transient private var path_ : Path = null
   @transient private var fs_ : FileSystem = null
 
+  protected override val customScopeName: Option[String] = Some(s"file stream [$id]")
```
Technically, this should/could be either `file stream` and `text file stream` depending on whether `ssc.fileStream` or `ssc.textFileStream` created it.  But I am okay as is.

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/dstream/DStream.scala:
```diff
@@ -111,6 +112,50 @@ abstract class DStream[T: ClassTag] (
+   * An optional custom name for all scopes generated by this DStream.
+   * If None, the name of the operation that created this DStream will be used.
+   */
+  protected val customScopeName: Option[String] = None
```
Lets not expose this to the public API. Rather do `private[streaming]` like `createSite` and `baseScope`

andrewor14 replied:
Wait, but protected is actually tighter than `private[streaming]`. This means only subclasses can access / override it

tdas replied:
But we dont want to expose it publicly. That's why may be its best to do `protected[streaming]`.

andrewor14 replied:
Do we not want users who create their own `DStreams` to be able to give their own names?

andrewor14 replied:
Ok, `protected[streaming]` it is

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/dstream/DStream.scala:
```diff
@@ -302,35 +347,22 @@ abstract class DStream[T: ClassTag] (
-          compute(time)
-        }
-        ssc.sparkContext.setCallSite(prevCallSite)
+        val newRDD = doCompute(time)
```
This should have `option` in the name as before.

tdas commented on external/twitter/src/main/scala/org/apache/spark/streaming/twitter/TwitterUtils.scala:
```diff
@@ -40,7 +40,7 @@ object TwitterUtils {
       filters: Seq[String] = Nil,
       storageLevel: StorageLevel = StorageLevel.MEMORY_AND_DISK_SER_2
-    ): ReceiverInputDStream[Status] = {
+    ): ReceiverInputDStream[Status] = ssc.withScope {
```
Why have this, when you are any ways overriding it with `customScopeName`?

tdas replied:
Wait, why at all have customScopeName in DStream, when you can always set the custom scope name using `withScope(someScopeName)`?

andrewor14 replied:
Answer to your first question: this is still need because you need the scope ID even if you don't need the name.

2nd question: It's to avoid duplicating the name of the stream. There are a few methods that create `*KafkaInputDStream` for instance.

tdas replied:
Well you are sort-of-duplicating some code/processing by adding `withScope` as well as `customScopeName`. Instead just add `withScope("kafka stream")` in one place and remove the extra `customeScopeName` from everywhere. And even though there are many versions of KafkaUtils.createStream, only one or two of them actually calls `new KafkaInputDStream`. So using `withScope("kafka stream")` in only those is sufficient. Also that more matches other ways of setting scope where the scope is defined by the caller method rather than by the entity itself. In fact, conceptually, an entity cannot really define the idea of scope around itself; only the caller/user of the entity can define the scope around the entity.

andrewor14 replied:
then how do we include the DStream IDs?

andrewor14 replied:
Alright, I implemented what you suggested in the latest commit. I had to add another helper method that `InputDStream` overrides to include the ID. For most streams there is not much duplication, but for `FileInputDStream`, for instance, we have to put `"file stream"` in 3 places. I think there are dis/advantages in both approaches. Let me know what you think about the latest one.

Commit:
Merge branch 'master' of github.com:apache/spark into dag-viz-streaming

Commit:
Review comments

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #32799 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32799/consoleFull) for   PR 6034 at commit [`074c00b`](https://github.com/apache/spark/commit/074c00bc72d0047a377bf677cb9ad30c1e2d16d3).

SparkQA commented:
[Test build #32799 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32799/consoleFull) for   PR 6034 at commit [`074c00b`](https://github.com/apache/spark/commit/074c00bc72d0047a377bf677cb9ad30c1e2d16d3).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32799/
Test FAILed.

Commit:
Fix style

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #32800 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32800/consoleFull) for   PR 6034 at commit [`1af0b0e`](https://github.com/apache/spark/commit/1af0b0e582cb6e78694728a8c589f3a70a1080a3).

Commit:
Pass in input stream name rather than defining it from within

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #32805 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32805/consoleFull) for   PR 6034 at commit [`6b36b0d`](https://github.com/apache/spark/commit/6b36b0da71f8c39214b42a0e102ca5a6e11d3669).

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #32806 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32806/consoleFull) for   PR 6034 at commit [`fa4e5fb`](https://github.com/apache/spark/commit/fa4e5fbb3aaca6fcfbdf358bf08371f0146b24a1).

Commit:
Remove a few unnecessary withScopes on aliases

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #32808 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32808/consoleFull) for   PR 6034 at commit [`0ca1801`](https://github.com/apache/spark/commit/0ca1801c10de10953b085a8438dd4f5c837d3007).

SparkQA commented:
[Test build #32800 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32800/consoleFull) for   PR 6034 at commit [`1af0b0e`](https://github.com/apache/spark/commit/1af0b0e582cb6e78694728a8c589f3a70a1080a3).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32800/
Test PASSed.

SparkQA commented:
[Test build #32806 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32806/consoleFull) for   PR 6034 at commit [`fa4e5fb`](https://github.com/apache/spark/commit/fa4e5fbb3aaca6fcfbdf358bf08371f0146b24a1).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32806/
Test FAILed.

SparkQA commented:
[Test build #32805 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32805/consoleFull) for   PR 6034 at commit [`6b36b0d`](https://github.com/apache/spark/commit/6b36b0da71f8c39214b42a0e102ca5a6e11d3669).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32805/
Test PASSed.

SparkQA commented:
[Test build #32808 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32808/consoleFull) for   PR 6034 at commit [`0ca1801`](https://github.com/apache/spark/commit/0ca1801c10de10953b085a8438dd4f5c837d3007).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32808/
Test PASSed.

tdas commented on external/mqtt/src/main/scala/org/apache/spark/streaming/mqtt/MQTTUtils.scala:
```diff
@@ -37,7 +37,7 @@ object MQTTUtils {
       topic: String,
       storageLevel: StorageLevel = StorageLevel.MEMORY_AND_DISK_SER_2
-    ): ReceiverInputDStream[String] = {
+    ): ReceiverInputDStream[String] = ssc.withNamedScope("MQTT stream") {
```
Make lower case.

Commit:
Merge branch 'master' of github.com:apache/spark into dag-viz-streaming

Commit:
Make MQTT lower case

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #32850 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32850/consoleFull) for   PR 6034 at commit [`fd07d22`](https://github.com/apache/spark/commit/fd07d22a2ace2b1e025ed19e2f311692ab02ccf1).

SparkQA commented:
[Test build #32850 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32850/consoleFull) for   PR 6034 at commit [`fd07d22`](https://github.com/apache/spark/commit/fd07d22a2ace2b1e025ed19e2f311692ab02ccf1).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32850/
Test FAILed.

Commit:
Merge branch 'master' of github.com:apache/spark into dag-viz-streaming

Commit:
Refactor DStream scope names again

Now it reads from the class name instead of us having to specify
it wherever we instantiate it. This reduces some more duplicate
code.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #32910 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32910/consoleFull) for   PR 6034 at commit [`1881802`](https://github.com/apache/spark/commit/188180227a5d8f0d677f2b9df4c5526f4024cbf7).

Commit:
Set scopes for foreachRDD properly

Previously, RDDs created inside the body of foreachRDD are not
scoped properly, such that low level Spark operations surface
to the UI. This is now fixed such that these RDDs are wrapped
in the `foreachRDD` scope.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #32911 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32911/consoleFull) for   PR 6034 at commit [`53b9936`](https://github.com/apache/spark/commit/53b993658eff0edf27117bd5ae43d3bc0588aee1).

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

Commit:
Fix tests

andrewor14 commented:
@tdas PTAL. I refactored the scope naming thing based on our offline discussions and fixed the `foreachRDD` case. It's ready for merge from my side.

SparkQA commented:
[Test build #814 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/814/consoleFull) for   PR 6034 at commit [`697c086`](https://github.com/apache/spark/commit/697c086043d12525a5bc9b7563d1ceebd4de6967).

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #32913 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32913/consoleFull) for   PR 6034 at commit [`697c086`](https://github.com/apache/spark/commit/697c086043d12525a5bc9b7563d1ceebd4de6967).

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32912/
Test FAILed.

SparkQA commented:
[Test build #32910 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32910/consoleFull) for   PR 6034 at commit [`1881802`](https://github.com/apache/spark/commit/188180227a5d8f0d677f2b9df4c5526f4024cbf7).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32910/
Test PASSed.

SparkQA commented:
[Test build #32911 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32911/consoleFull) for   PR 6034 at commit [`53b9936`](https://github.com/apache/spark/commit/53b993658eff0edf27117bd5ae43d3bc0588aee1).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32911/
Test FAILed.

andrewor14 commented:
retest this please

SparkQA commented:
[Test build #815 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/815/consoleFull) for   PR 6034 at commit [`697c086`](https://github.com/apache/spark/commit/697c086043d12525a5bc9b7563d1ceebd4de6967).

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #32918 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32918/consoleFull) for   PR 6034 at commit [`697c086`](https://github.com/apache/spark/commit/697c086043d12525a5bc9b7563d1ceebd4de6967).

SparkQA commented:
[Test build #814 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/814/consoleFull) for   PR 6034 at commit [`697c086`](https://github.com/apache/spark/commit/697c086043d12525a5bc9b7563d1ceebd4de6967).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public class JavaOneVsRestExample `
  * `      head("OneVsRest Example: multiclass to binary reduction using OneVsRest")`
  * `class DataTypeSingleton(type):`
  * `class NullType(DataType):`
  * `class AtomicType(DataType):`
  * `class NumericType(AtomicType):`
  * `class IntegralType(NumericType):`
  * `class FractionalType(NumericType):`
  * `class StringType(AtomicType):`
  * `class BinaryType(AtomicType):`
  * `class BooleanType(AtomicType):`
  * `class DateType(AtomicType):`
  * `class TimestampType(AtomicType):`
  * `class DecimalType(FractionalType):`
  * `class DoubleType(FractionalType):`
  * `class FloatType(FractionalType):`
  * `class ByteType(IntegralType):`
  * `class IntegerType(IntegralType):`
  * `class LongType(IntegralType):`
  * `class ShortType(IntegralType):`
  * `class Column(object):`
  * `class GroupedData(object):`

SparkQA commented:
[Test build #32913 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32913/consoleFull) for   PR 6034 at commit [`697c086`](https://github.com/apache/spark/commit/697c086043d12525a5bc9b7563d1ceebd4de6967).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32913/
Test PASSed.

SparkQA commented:
[Test build #815 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/815/consoleFull) for   PR 6034 at commit [`697c086`](https://github.com/apache/spark/commit/697c086043d12525a5bc9b7563d1ceebd4de6967).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
[Test build #32918 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32918/consoleFull) for   PR 6034 at commit [`697c086`](https://github.com/apache/spark/commit/697c086043d12525a5bc9b7563d1ceebd4de6967).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32918/
Test FAILed.

SparkQA commented:
[Test build #816 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/816/consoleFull) for   PR 6034 at commit [`697c086`](https://github.com/apache/spark/commit/697c086043d12525a5bc9b7563d1ceebd4de6967).

tdas commented on external/flume/src/main/scala/org/apache/spark/streaming/flume/FlumeUtils.scala:
```diff
@@ -60,10 +60,7 @@ object FlumeUtils {
       storageLevel: StorageLevel,
       enableDecompression: Boolean
     ): ReceiverInputDStream[SparkFlumeEvent] = {
-    val inputStream = new FlumeInputDStream[SparkFlumeEvent](
```
No need to change this file any more. isnt it?

andrewor14 replied:
yes, but we're literally putting it in a val here and just returning it...

tdas commented on external/twitter/src/main/scala/org/apache/spark/streaming/twitter/TwitterUtils.scala:
```diff
@@ -21,8 +21,8 @@ import twitter4j.Status
 import org.apache.spark.storage.StorageLevel
 import org.apache.spark.streaming.StreamingContext
-import org.apache.spark.streaming.api.java.{JavaReceiverInputDStream, JavaDStream, JavaStreamingContext}
-import org.apache.spark.streaming.dstream.{ReceiverInputDStream, DStream}
```
Lets not change this file with this patch.

tdas commented on external/mqtt/src/main/scala/org/apache/spark/streaming/mqtt/MQTTUtils.scala:
```diff
@@ -21,8 +21,8 @@ import scala.reflect.ClassTag
 import org.apache.spark.storage.StorageLevel
 import org.apache.spark.streaming.StreamingContext
-import org.apache.spark.streaming.api.java.{JavaReceiverInputDStream, JavaStreamingContext, JavaDStream}
-import org.apache.spark.streaming.dstream.{ReceiverInputDStream, DStream}
```
Lets not change this file with this patch since not needed.

andrewor14 replied:
alright, I will add these unused imports back...

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/dstream/DStream.scala:
```diff
@@ -338,6 +369,41 @@ abstract class DStream[T: ClassTag] (
+   * Wrap a body of code such that the call site and operation scope
+   * information are passed to the RDDs created in this body properly.
+   */
+  protected def createRDDWith[U](time: Time)(body: => U): U = {
```
the name is incomplete. `createRDDWith` does not make sense. It has to `withSomething`. Maybe `withEnv`?
Also why did you change it from `doCompute`? Is there are point in wrapping whole of `getOrCompute` in this function, rather than just the `compute()` function as you were doing it earlier with `doCompute`?

andrewor14 replied:
This method is hard to name, but I don't think `withEnv` is a descriptive name at all. If you can come up with others, I am open to changing it.

I put it in its own method because we need to reuse it in `ForEachDStream`

andrewor14 replied:
As discussed offline, we will rename this to `createRDDWithLocalProperties`

SparkQA commented:
[Test build #816 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/816/consoleFull) for   PR 6034 at commit [`697c086`](https://github.com/apache/spark/commit/697c086043d12525a5bc9b7563d1ceebd4de6967).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

Commit:
Review feedback

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #32933 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32933/consoleFull) for   PR 6034 at commit [`84d0656`](https://github.com/apache/spark/commit/84d06565f4cb6f27fda295d54230ba5f7c1d6068).

SparkQA commented:
[Test build #32933 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32933/consoleFull) for   PR 6034 at commit [`84d0656`](https://github.com/apache/spark/commit/84d06565f4cb6f27fda295d54230ba5f7c1d6068).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32933/
Test PASSed.

Commit:
Rename createRDDWith

Commit:
Merge branch 'master' of github.com:apache/spark into dag-viz-streaming

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #33002 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33002/consoleFull) for   PR 6034 at commit [`932a64a`](https://github.com/apache/spark/commit/932a64a3c10623c58e9cc3b2a99e2b5f1e8c74a4).

tdas commented:
LGTM! Thanks @andrewor14 for this patch.

SparkQA commented:
[Test build #33002 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33002/consoleFull) for   PR 6034 at commit [`932a64a`](https://github.com/apache/spark/commit/932a64a3c10623c58e9cc3b2a99e2b5f1e8c74a4).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33002/
Test FAILed.

andrewor14 commented:
retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #33009 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33009/consoleFull) for   PR 6034 at commit [`932a64a`](https://github.com/apache/spark/commit/932a64a3c10623c58e9cc3b2a99e2b5f1e8c74a4).

SparkQA commented:
[Test build #33009 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33009/consoleFull) for   PR 6034 at commit [`932a64a`](https://github.com/apache/spark/commit/932a64a3c10623c58e9cc3b2a99e2b5f1e8c74a4).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33009/
Test PASSed.

andrewor14 commented:
Great, merging into master 1.4.
